lightning/src/routing/scoring.rs

```diff
@@ -1228,119 +1232,102 @@ impl<L: Deref<Target = u64>, BRT: Deref<Target = HistoricalBucketRangeTracker>,
 			log_debug!(logger, "Setting max liquidity of {} from {} to {}", chan_descr, existing_max_msat, amount_msat);
-			self.set_max_liquidity_msat(amount_msat);
+			self.set_max_liquidity_msat(amount_msat, duration_since_epoch);
 		} else {
 			log_trace!(logger, "Max liquidity of {} is {} (already less than or equal to {})",
 				chan_descr, existing_max_msat, amount_msat);
 		}
-		self.update_history_buckets(0);
+		self.update_history_buckets(0, duration_since_epoch);
 	}
 
 	/// Adjusts the channel liquidity balance bounds when failing to route `amount_msat` downstream.
-	fn failed_downstream<Log: Deref>(&mut self, amount_msat: u64, chan_descr: fmt::Arguments, logger: &Log) where Log::Target: Logger {
+	fn failed_downstream<Log: Deref>(
+		&mut self, amount_msat: u64, duration_since_epoch: Duration, chan_descr: fmt::Arguments, logger: &Log
+	) where Log::Target: Logger {
 		let existing_min_msat = self.min_liquidity_msat();
 		if amount_msat > existing_min_msat {
 			log_debug!(logger, "Setting min liquidity of {} from {} to {}", existing_min_msat, chan_descr, amount_msat);
-			self.set_min_liquidity_msat(amount_msat);
+			self.set_min_liquidity_msat(amount_msat, duration_since_epoch);
 		} else {
 			log_trace!(logger, "Min liquidity of {} is {} (already greater than or equal to {})",
 				chan_descr, existing_min_msat, amount_msat);
 		}
-		self.update_history_buckets(0);
+		self.update_history_buckets(0, duration_since_epoch);
 	}
 
 	/// Adjusts the channel liquidity balance bounds when successfully routing `amount_msat`.
-	fn successful<Log: Deref>(&mut self, amount_msat: u64, chan_descr: fmt::Arguments, logger: &Log) where Log::Target: Logger {
+	fn successful<Log: Deref>(&mut self,
+		amount_msat: u64, duration_since_epoch: Duration, chan_descr: fmt::Arguments, logger: &Log
+	) where Log::Target: Logger {
 		let max_liquidity_msat = self.max_liquidity_msat().checked_sub(amount_msat).unwrap_or(0);
 		log_debug!(logger, "Subtracting {} from max liquidity of {} (setting it to {})", amount_msat, chan_descr, max_liquidity_msat);
-		self.set_max_liquidity_msat(max_liquidity_msat);
-		self.update_history_buckets(amount_msat);
+		self.set_max_liquidity_msat(max_liquidity_msat, duration_since_epoch);
+		self.update_history_buckets(amount_msat, duration_since_epoch);
 	}
 
 	/// Updates the history buckets for this channel. Because the history buckets track what we now
 	/// know about the channel's state *prior to our payment* (i.e. what we assume is "steady
 	/// state"), we allow the caller to set an offset applied to our liquidity bounds which
 	/// represents the amount of the successful payment we just made.
-	fn update_history_buckets(&mut self, bucket_offset_msat: u64) {
-		let half_lives = self.now.duration_since(*self.last_updated).as_secs()
+	fn update_history_buckets(&mut self, bucket_offset_msat: u64, duration_since_epoch: Duration) {
+		let half_lives =
+			duration_since_epoch.checked_sub(*self.offset_history_last_updated)
```

Comment:
This isn't accurate given that `offset_history_last_updated` is updated in `set_min_liquidity_msat` and `set_max_liquidity_msat` which could (but may not) be called prior to calling `update_history_buckets`. Do we have tests to catch this?